### PR TITLE
inference: refine `setglobal!` rt for invalid assignments

### PR DESCRIPTION
**Pull request by aviatesk**
@nanosoldier `runbenchmarks("inference", vs=":master")`

### CODE REVIEW

File: Compiler/src/abstractinterpretation.jl [hunk 1]
```diff
@@ -3396,28 +3396,31 @@ function abstract_eval_globalref(interp::AbstractInterpreter, g::GlobalRef, sv::
...
     end
-
     ty = partition_restriction(partition)
-    if !(widenconst(newty) <: ty)
-        return TypeError
+    wnewty = widenconst(newty)
+    if !hasintersect(wnewty, ty)
+        return Pair{Any,Any}(Bottom, TypeError)
+    elseif !(wnewty <: ty)
+        return Pair{Any,Any}(newty, TypeError)
```
Comment (vtjnash) [inline Compiler/src/abstractinterpretation.jl:3416]: use the typeintersect_tfunc here to compute newty most precisely?

**Review by vtjnash — APPROVED** [review]

File: Compiler/src/abstractinterpretation.jl (hunk 1 above)
Comment (aviatesk) [reply Compiler/src/abstractinterpretation.jl:3416]: Do you mean `tmeet(𝕃, newty, ty)`?

Comment (vtjnash) [reply Compiler/src/abstractinterpretation.jl:3416]: Yes, I guess that is true. Often the types are in the lifted domain, but in this case the partition_restriction is already in the value domain, so the instanceof_tfunc isn't necessary

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/56622) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/94ece59_vs_bdd4e05/report.md).

**Review by Keno — APPROVED** [review]
LGTM modulo Jameson's comment

**Pull request merged**